### PR TITLE
UCT/CONFIG: Added support for out-of-order data placement for rc and dc

### PR DESCRIPTION
**Pull request by alinask**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3924/ for details (Mellanox internal link).

**Comment by alinask** [comment]
@yosefe can you please review?
@MattBBaker was ORNL jenkins not triggered?

File: src/uct/ib/dc/base/dc_iface.c [hunk 1]
```diff
@@ -14,7 +14,11 @@ const static char *uct_dc_tx_policy_names[] = {
 };
 
 ucs_config_field_t uct_dc_iface_config_table[] = {
-    {"RC_", "IB_TX_QUEUE_LEN=128;FC_ENABLE=n", NULL,
+    {"RC_", "IB_TX_QUEUE_LEN=128;FC_ENABLE=n;"
+#if HAVE_DECL_IBV_EXP_QP_OOO_RW_DATA_PLACEMENT
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.c]: why needed? default is no anyway?

File: src/uct/ib/base/ib_verbs.h [hunk 2]
```diff
@@ -113,6 +113,17 @@
 #  define IBV_DEVICE_HAS_NOP(_attr)                 0
 #endif /* HAVE_DECL_IBV_EXP_WR_NOP */
 
+/*
+ * Adaptive Routing support
+ */
+#if HAVE_DECL_IBV_EXP_QP_OOO_RW_DATA_PLACEMENT
+#  define UCX_IB_DEV_IS_OOO_SUPPORTED(_ibdev, _transport)  \
```
Comment (yosefe) [inline src/uct/ib/base/ib_verbs.h]: let's make this macro accept _attr and not _ibdev like rest of the macros in this file

File: src/uct/ib/rc/base/rc_iface.c [hunk 3]
```diff
@@ -667,7 +674,12 @@ ucs_status_t uct_rc_iface_qp_connect(uct_rc_iface_t *iface, struct ibv_qp *qp,
                                      const uint32_t dest_qp_num,
                                      struct ibv_ah_attr *ah_attr)
 {
+#if HAVE_DECL_IBV_EXP_QP_OOO_RW_DATA_PLACEMENT
+    struct ibv_exp_qp_attr qp_attr = {0};
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.c]: why need to init to {0}, line 685 already does memset

File: src/uct/ib/rc/base/rc_iface.h [hunk 4]
```diff
@@ -204,6 +205,8 @@ struct uct_rc_iface {
         uint8_t              retry_cnt;
         uint8_t              max_rd_atomic;
         enum ibv_mtu         path_mtu;
+        /* Enable out-of-order RDMA data placement */
+        unsigned             ooo_rw;
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.h]: can be uint8_t

File: src/uct/ib/rc/base/rc_iface.h [hunk 5]
```diff
@@ -128,6 +128,7 @@ struct uct_rc_iface_config {
     uct_ib_iface_config_t    super;
     uct_ib_mtu_t             path_mtu;
     unsigned                 max_rd_atomic;
+    unsigned                 ooo_rw; /* Enable out-of-order RDMA data placement */
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.h]: should be int (boolean config is int)

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3932/ for details (Mellanox internal link).

**Comment by alinask** [comment]
@yosefe please review

**Pull request merged**